[lldb][Mach-O] Read dyld_all_image_infos addr from `main bin spec` LC_NOTE

jasonmolenda opened the pull request:
Mach-O corefiles have LC_NOTE metadata, one LC_NOTE that lldb recognizes is `main bin spec` which can specify that this is a kernel corefile, userland corefile, or firmware/standalone corefile. With a userland corefile, the LC_NOTE would specify the virtual address of the dyld binary's Mach-O header.  lldb would create a Module from that in-memory binary, find the `dyld_all_image_infos` object in dyld's DATA segment, and use that object to find all of the binaries present in the corefile.

ProcessMachCore takes the metadata from this LC_NOTE and passes the address to the DynamicLoader plugin via its `GetImageInfoAddress()` method, so the DynamicLoader can find all of the binaries and load them in the Target at their correct virtual addresses.

We have a corefile creator who would prefer to specify the address of `dyld_all_image_infos` directly, instead of specifying the address of dyld and parsing that to find the object.  DynamicLoaderMacOSX, the DynamicLoader plugin being used here, will accept either a dyld virtual address or a `dyld_all_image_infos` virtual address from ProcessMachCore, and do the correct thing with either value.

lldb's process save-core mach-o corefile reader will continue to specify the virtual address of the dyld binary.

rdar://144322688

jasonmolenda committed:
[lldb][Mach-O] Read dyld_all_image_infos addr from `main bin spec` LC_NOTE

Mach-O corefiles have LC_NOTE metadata, one LC_NOTE that lldb
recognizes is `main bin spec` which can specify that this is a
kernel corefile, userland corefile, or firmware/standalone corefile.
With a userland corefile, the LC_NOTE would specify the virtual
address of the dyld binary's Mach-O header.  lldb would create a
Module from that in-memory binary, find the `dyld_all_image_infos`
object in dyld's DATA segment, and use that object to find all of
the binaries present in the corefile.

ProcessMachCore takes the metadata from this LC_NOTE and passes the
address to the DynamicLoader plugin via its `GetImageInfoAddress()`
method, so the DynamicLoader can find all of the binaries and load
them in the Target at their correct virtual addresses.

We have a corefile creator who would prefer to specify the address
of `dyld_all_image_infos` directly, instead of specifying the address
of dyld and parsing that to find the object.  DynamicLoaderMacOSX,
the DynamicLoader plugin being used here, will accept either a
dyld virtual address or a `dyld_all_image_infos` virtual address
from ProcessMachCore, and do the correct thing with either value.

lldb's process save-core mach-o corefile reader will continue to
specify the virtual address of the dyld binary.

rdar://144322688

jasonmolenda commented:
I wanted to add a new `eBinaryType` enum from the ObjectFile to describe what address is being returned.  But within ProcessMachCore, there was no need to add the distinction between the existing `m_dyld_addr` and the new `m_dyld_all_image_infos_addr`, I did it to keep the meaning of the address clear at this layer, and to make logging more clear.  But ProcessMachCore returns either address to DynamicLoaderMacOSX, and it sniffs for a Mach-O magic number in the first 4 bytes to disambiguate between them already.  I could have ProcessMachCore put either value in `m_dyld_addr` and everything would still work.

JDevlieghere commented on lldb/include/lldb/Symbol/ObjectFile.h:
```diff
@@ -81,9 +81,12 @@ class ObjectFile : public std::enable_shared_from_this<ObjectFile>,
-    eBinaryTypeKernel,    /// kernel binary
-    eBinaryTypeUser,      /// user process binary
-    eBinaryTypeStandalone /// standalone binary / firmware
+    eBinaryTypeKernel,            /// kernel binary
+    eBinaryTypeUser,              /// user process binary,
+                                  /// dyld addr
+    eBinaryTypeUserAllImageInfos, /// user process binary,
+                                  /// dyld_all_image_infos addr
+    eBinaryTypeStandalone         /// standalone binary / firmware
```
```suggestion
    /// kernel binary
    eBinaryTypeKernel, 
    /// user process binary, dyld addr
    eBinaryTypeUser,        
    /// user process binary, dyld_all_image_infos addr
    eBinaryTypeUserAllImageInfos, 
    /// standalone binary / firmware
    eBinaryTypeStandalone         
```

Doxygen requires `///<` for inline comments like this. I don't know how that works with multiline commetns and given our 80 col limit, documenting them above the value is more readable anyway.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/Mach-O/ObjectFileMachO.cpp:
```diff
@@ -5669,6 +5673,10 @@ bool ObjectFileMachO::GetCorefileMainBinaryInfo(addr_t &value,
+          case 4:
+            type = eBinaryTypeUserAllImageInfos;
+            typestr = "userland dyld_all_image_infos";
+            break;
```
Should this have a default case that sets `type` to `eBinaryTypeInvalid`?

bulbazord replied:
If you do, it might be useful to add a log. If that situation were to happen, you'd see it in the logs immediately and know that either the core file looks unexpected or LLDB has a bug in core file handling.

jasonmolenda replied:
The `typestr` is set to `"unrecognized type"` and will be printed if logging is enabled, but the `type` enum is uninitialized and will also be printed, which could be confusing.  I'll add an initialization.  This method doesn't currently take a Target or Process pointer so I can't print a message to the user asynchronously, it'll require them to enable logging when something is going wrong to find it.

JDevlieghere reviewed (approved):
LGTM

jasonmolenda committed:
Update lldb/include/lldb/Symbol/ObjectFile.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

jasonmolenda committed:
Initialize `type` to a known invalid value so
the logging message is clearer when an unknown
type is found in the main bin spec.

jasonmolenda committed:
whitespace fix
